File: vs2003_cd01/Program Files/Microsoft Visual Studio .NET 2003/SDK/v1.1/Samples/Technologies/Remoting/Basic/RemotingHello/Client/consoleclientdirect/HelloAsync2.cs
/*=====================================================================

  File:      HelloAsync2.cs

---------------------------------------------------------------------
This file is part of the Microsoft .NET Framework SDK Code Samples.

  Copyright (C) Microsoft Corporation.  All rights reserved.

This source code is intended only as a supplement to Microsoft
Development Tools and/or on-line documentation.  See these other
materials for detailed information regarding Microsoft code samples.
 
THIS CODE AND INFORMATION ARE PROVIDED "AS IS" WITHOUT WARRANTY OF ANY
KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
PARTICULAR PURPOSE.
=====================================================================*/

using System;
using System.Runtime.Remoting;
using System.Runtime.Remoting.Messaging;
using System.Threading;

using Hello;

public class HelloAsync2
{
    public static void Main(String[] args)
    {
        Console.WriteLine(".NET Remoting Sample - HelloAsync2");
        Console.WriteLine("----------------------------------\n");

        HelloAsync2 helloAsync = new HelloAsync2();
        helloAsync.Run();
    }

    public void Run()
    {
        String name = "Bill";
        String configFilename = "HelloAsync2.exe.config";

        Console.WriteLine("Configuring Remoting from {0}", configFilename);
        RemotingConfiguration.Configure(configFilename);

        Console.WriteLine("Obtaining Proxy for HelloService, using new");
        HelloService helloService = new HelloService();

        Console.WriteLine("Creating ReturnGreetingDelegate to HelloService.ReturnGreeting");
        ReturnGreetingDelegate returnGreetingDelegate = new ReturnGreetingDelegate(helloService.ReturnGreeting);

        Console.WriteLine("Creating AsyncCallback to HelloAsync.greettingCallback");
        AsyncCallback asyncCallback = new AsyncCallback(this.greetingCallBack);

        Console.WriteLine("Creating ManualResetEvent, will be fired when greetingCallBack is complete");
        ManualResetEvent evt = new ManualResetEvent(false);

        Console.WriteLine("ManualResetEvent is passed in async state object, greetingCallBack will use it");
        Object stateObject = evt;

        Console.WriteLine("Calling returnGreetingDelegate.BeginInvokeInvoke");
        IAsyncResult ar = returnGreetingDelegate.BeginInvoke(name, asyncCallback, stateObject);

        Console.WriteLine("Waiting for greetingCallback to completed");
        evt.WaitOne();
        Console.WriteLine("Event fired");
    }

    public delegate String ReturnGreetingDelegate(String name);

    public void greetingCallBack(IAsyncResult ar)
    {
        ReturnGreetingDelegate returnGreetingDelegate = (ReturnGreetingDelegate)((AsyncResult)ar).AsyncDelegate;

        Console.WriteLine("Calling returnGreetingDelegate.EndInvoke");
        String greeting = returnGreetingDelegate.EndInvoke(ar);

        Console.WriteLine("returnGreetingDelegate.EndInvoke returned: {0}", greeting);

        Console.WriteLine("Fire event that greetingCallBack is complete");
        ManualResetEvent stateObjectEvt = (ManualResetEvent)ar.AsyncState;
        stateObjectEvt.Set();
    }
}

File: vs2003_cd01/Program Files/Microsoft Visual Studio .NET 2003/SDK/v1.1/Samples/Technologies/Remoting/VB/Basic/RemotingHello/Client/ConsoleClientAll/HelloAsync2.vb

Imports System
Imports System.Runtime.Remoting
Imports System.Runtime.Remoting.Messaging
Imports System.Threading
Imports Microsoft.VisualBasic

Imports Hello



Public Class HelloAsync2
    
    Public Shared Sub Main()
        Console.WriteLine(".NET Remoting Sample - HelloAsync2")
        Console.WriteLine("----------------------------------" + ControlChars.Cr)
        
        Dim helloAsync As New HelloAsync2()
        helloAsync.Run()
    End Sub 'Main
    
    
    Public Sub Run()
        Dim name As String = "Bill"
        Dim configFilename As String = "HelloAsync2.exe.config"
        
        Console.WriteLine("Configuring Remoting from {0}", configFilename)
        RemotingConfiguration.Configure(configFilename)
        
        Console.WriteLine("Obtaining Proxy for HelloService, using new")
        Dim helloService As New HelloService()
        
        Console.WriteLine("Creating ReturnGreetingDelegate to HelloService.ReturnGreeting")
        Dim returnGreetingDelegate As New ReturnGreetingDelegate(AddressOf helloService.ReturnGreeting)
        
        Console.WriteLine("Creating AsyncCallback to HelloAsync.greettingCallback")
        Dim asyncCallback As New AsyncCallback(AddressOf Me.greetingCallBack)
        
        Console.WriteLine("Creating ManualResetEvent, will be fired when greetingCallBack is complete")
        Dim evt As New ManualResetEvent(False)
        
        Console.WriteLine("ManualResetEvent is passed in async state object, greetingCallBack will use it")
        Dim stateObject As [Object] = evt
        
        Console.WriteLine("Calling returnGreetingDelegate.BeginInvokeInvoke")
        Dim ar As IAsyncResult = returnGreetingDelegate.BeginInvoke(name, asyncCallback, stateObject)
        
        Console.WriteLine("Waiting for greetingCallback to completed")
        evt.WaitOne()
        Console.WriteLine("Event fired")
    End Sub 'Run
    
    
    Delegate Function ReturnGreetingDelegate(name As String) As String
    
    
    Public Sub greetingCallBack(ar As IAsyncResult)
        Dim returnGreetingDelegate As ReturnGreetingDelegate = CType(CType(ar, AsyncResult).AsyncDelegate, ReturnGreetingDelegate)
        
        Console.WriteLine("Calling returnGreetingDelegate.EndInvoke")
        Dim greeting As String = returnGreetingDelegate.EndInvoke(ar)
        
        Console.WriteLine("returnGreetingDelegate.EndInvoke returned: {0}", greeting)
        
        Console.WriteLine("Fire event that greetingCallBack is complete")
        Dim stateObjectEvt As ManualResetEvent = CType(ar.AsyncState, ManualResetEvent)
        stateObjectEvt.Set()
    End Sub 'greetingCallBack
End Class 'HelloAsync2



